### PR TITLE
Add capability to use ros2 param set for array types

### LINKED ISSUE #198: [ros2param] ros2 param set does not work for array types
## Bug report

- Operating System:
  - Arch Linux 4.20.8-arch1-1-ARCH
- Installation type:
  - source
- Version or commit hash:
  - 66efd10c61bd54ebdff0035102b9878a2acf5b4b
- DDS implementation:
  - Fast-RTPS

#### Steps to reproduce issue
*rclpy*:

- Create a node with a parameter of an array type, such as integer array:
```
import rclpy
from rclpy.parameter import Parameter


def main(args=None):
    rclpy.init(args=args)

    node = rclpy.create_node('param_array')
    node.set_parameters([Parameter("foo", Parameter.Type.INTEGER_ARRAY, [0, 0])])
    
    def timer_callback():
        node.get_logger().info('Value: {}, type: {}'.format(
            node.get_parameter("foo").value,
            type(node.get_parameter("foo").value)))

    timer = node.create_timer(1, timer_callback)
    rclpy.spin(node)

    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
```
- Set parameter using `ros2 param set`:
```
ros2 param set /param_array foo "[1, 1]"
```

*rclcpp:*
- Create a node with a parameter of an array type, such as integer array:
```
std::vector<int64_t> foo;
get_parameter("foo", foo);
```
- Set parameter using `ros2 param set`

#### Expected behavior

Array parameter set successfully.

#### Actual behavior

*rclpy*:
No exception, but parameter is set to the raw string value:
```
[INFO] [param_array]: Value: [0, 0], type: <class 'list'>
[INFO] [param_array]: Value: [0, 0], type: <class 'list'>
# param set is run
[INFO] [param_array]: Value: [1, 1], type: <class 'str'>
[INFO] [param_array]: Value: [1, 1], type: <class 'str'>
[INFO] [param_array]: Value: [1, 1], type: <class 'str'>
```

*rclcpp*:
ParameterTypeException thrown:
```
terminate called after throwing an instance of 'rclcpp::ParameterTypeException'                                     
  what():  expected [integer_array] got [string]
```

### PR DESCRIPTION
**Pull request by sgvandijk**
Using YAML parsing to determine the value parameter type, the supported parameter types are extended to array types. Only the byte array type is excluded, I don't see a way of expressing that in YAML in an unambiguous way.

Fixes #198

### CODE REVIEW

File: ros2param/ros2param/api/__init__.py [hunk 1]
```diff
@@ -17,20 +17,44 @@
...
+        return value
+
+    if isinstance(yaml_value, bool):
         value.type = ParameterType.PARAMETER_BOOL
-        value.bool_value = string_value.lower() == 'true'
-    elif _is_integer(string_value):
+        value.bool_value = yaml_value
+    elif isinstance(yaml_value, int):
         value.type = ParameterType.PARAMETER_INTEGER
         value.integer_value = int(string_value)
```
Comment (dirk-thomas) [inline ros2param/ros2param/api/__init__.py]: This could assign `yaml_value` instead which makes the `int()` call is obsolete.

Same for `float` below.

**Comment by sgvandijk** [comment]
Thanks for the review! I've applied the requested changes. I've had a look at the failing build, looks not related to this:

ERROR: unable to process source [https://raw.githubusercontent.com/ros/rosdistro/master/releases/fuerte.yaml]

**Comment by cottsay** [comment]
That appears to have been an infrastructure issue with Github that has since been resolved. It was also worked around by [ros-infrastructure/rosdep#663](https://github.com/ros-infrastructure/rosdep/pull/663).

I'll try to re-trigger the build.

File: ros2param/test/test_api.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from rcl_interfaces.msg import ParameterType
+from ros2param.api import get_parameter_value
+
+
+def test_bool():
```
Comment (juanrh) [inline ros2param/test/test_api.py]: All these tests are quite similar, so we could merge them in a single parametrized test, that is easier to maintain. Something like 

```python 

import pytest

@pytest.mark.parametrize("string_value,value_attribute,expected_type,expected_value", [
    ('true', 'bool_value', ParameterType.PARAMETER_BOOL, True),
    ('foo', 'string_value', ParameterType.PARAMETER_STRING, 'foo'),
   ('["foo", "bar", "buzz"]', 'string_array_value', ParameterType.PARAMETER_STRING_ARRAY, ['foo', 'bar', 'buzz'])
])
def test_get_parameter_value(string_value, value_attribute, expected_type, expected_value):
    value = get_parameter_value(string_value=string_value)
    assert value.type == expected_type
    assert getattr(value, value_attribute) is expected_value
```

Comment (sgvandijk) [reply ros2param/test/test_api.py]: I like that, doing that now.

**Comment by sgvandijk** [comment]
Just to check if there are any more blockers for this PR?

**Comment by dirk-thomas** [comment]
@sgvandijk Sorry for the late reply. I applied some minor cleanup:
* I removed two obsolete functions in 59c2052e923400bfd5d34fec014318f019e2c4e3
* I updated the test to expect the numeric lists to be of type `array.array` (which is due to a recent change in the Python mapping, see ros2/rosidl_python#35): 492fba03a705cd5a2362837fc8e10444418ac889

CI is looking good: [Build Status](https://ci.ros2.org/job/ci_linux/6789/)

**Comment by dirk-thomas** [comment]
Since I made the changes through the web interface the commits were not signed of making the DCO check fail. I took the liberty to force push my last two commits with a proper "Signed-off-by" line.

This is good to be :boat: now. Thank you for the contribution.

**Pull request merged**